### PR TITLE
Implement a txs syncer using electrum WS & batched JSONRPC calls

### PR DESCRIPTION
**Pull request by tiero**
Logic is not finished, it's based on Electrum GDK code https://github.com/Blockstream/gdk/blob/master/subprojects/gdk_rust/gdk_electrum/src/lib.rs#L1346-L1511

### CODE REVIEW

File: src/background.ts [hunk 1]
```diff
@@ -18,160 +17,196 @@ MainAccount m/84'/1776'/0'
...
+        } else {
+          lastUsed.external = max + batchCount * GAP_LIMIT;
+        }
+      }
+
+
+      let flattened: GetHistoryResponse[] = histories.flat();
+      console.log(`${i}/batch(${batchCount}) ${flattened.length}`);
+
+      if (flattened.length === 0) {
```
Comment (tiero) [inline src/background.ts:92]: this is interesting: we stop looping when the flattened arrays contains 0 elements ie. empty gap limit

File: src/background.ts [hunk 2]
```diff
@@ -18,160 +17,196 @@ MainAccount m/84'/1776'/0'
...
+  static BASE_DERIVATIONT_PATH_TESTNET = "m/84'/1'/0'";
+
+
+  constructor(node: BIP32Interface, private network: networks.Network, private baseDerivationPath: string = Account.BASE_DERIVATION_PATH_LEGACY) {
+    this.node = node.derivePath(baseDerivationPath);
+    this.cache = {};
+  }
+
+  // Derive a range from start to end index of public keys applying the base derivation path
+  deriveBatch(start: number, end: number, isInternal: boolean): string[] {
```
Comment (tiero) [inline src/background.ts:131]: this should be based on [this](https://github.com/Blockstream/gdk/blob/9fbc7abbea98f33f698b9be9da10a452860a7b94/subprojects/gdk_rust/gdk_electrum/src/account.rs#L867-L888), but I don't fully understand the code, especially they multiply start * BATCH_SIZE etc.. therefore I changed a bit

Pretty sure there is a reason, as not all the code as been ported

File: src/syncer.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        } else {
+          lastUsed.external = max + batchCount * GAP_LIMIT;
+        }
+      }
+
+
+      let flattened: GetHistoryResponse[] = histories.flat();
+      console.log(`${i}/batch(${batchCount}) ${flattened.length}`);
+
+      if (flattened.length === 0) {
```
Comment (tiero) [inline src/syncer.ts]: this is the interesting trick: we stop looping when the flattened arrays contain 0 elements ie. an empty gap limit

File: src/account.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  static BASE_DERIVATION_PATH_LEGACY = "m/84'/0'/0'";
+  static BASE_DERIVATIONT_PATH_TESTNET = "m/84'/1'/0'";
+
+
+  constructor(node: BIP32Interface, private network: networks.Network, private baseDerivationPath: string = Account.BASE_DERIVATION_PATH_LEGACY) {
+    this.node = node.derivePath(baseDerivationPath);
+    this.cache = {};
+  }
+
+  // Derive a range from start to end index of public keys applying the base derivation path
```
Comment (tiero) [inline src/account.ts:40]: this should be based on [this](https://github.com/Blockstream/gdk/blob/9fbc7abbea98f33f698b9be9da10a452860a7b94/subprojects/gdk_rust/gdk_electrum/src/account.rs#L867-L888), but I don't fully understand the code, especially they multiply start * BATCH_SIZE etc.. therefore I changed a bit

Pretty sure there is a reason, as not all the code as been ported

**Comment by tiero** [comment]
Notice: batching requests seem to now work at the moment, ie. socket closes the connection

**Comment by louisinger** [comment]
I'm closing this, I've extended the PR here: #2

**Pull request closed without merging**